OpenRA.Game/Map/CellLayer.cs

```diff
@@ -45,9 +45,23 @@ public void CopyValuesFrom(CellLayer<T> anotherLayer)
 			Array.Copy(anotherLayer.entries, entries, entries.Length);
 		}
 
+		public static CellLayer<T> CreatePrototype(Func<int, int, T> creator, int mapWidth, int mapHeight, TileShape tileShape)
+		{
+			var size = new Size(mapWidth, mapHeight);
+			var defaultCellInfoLayer = new CellLayer<T>(tileShape, size);
+			for (var v = 0; v < mapHeight; v++)
+				for (var u = 0; u < mapWidth; u++)
+					defaultCellInfoLayer[new MPos(u, v)] = creator(u, v);
+
+			return defaultCellInfoLayer;
+		}
+
 		// Resolve an array index from cell coordinates
 		int Index(CPos cell)
 		{
+			if (Shape == TileShape.Rectangle)
+				return cell.Y * Size.Width + cell.X;
```

Comment:
For the time being, I'll leave it back to where it was because couldn't come up with another way to refactor and reuse the calculation in an elegant way, I'll study again how that will impact performance.